File: Workshop 3 Solsystemet-20230210/Planet.m
classdef Planet <handle
    %PLANET Class. Based on the method of https://ssd.jpl.nasa.gov/txt/aprx_pos_planets.pdf

    properties
        name
        radius
        coordinates=[0;0];
        color
        resolution=200
        a           %halve storakse
        da          %ændring i halve storakse pr. århundrede
        e           %excentricitet
        de          %ændring i excentricitet pr. århundrede
        L           %middelbreddegrad
        dL          %ændring i middelbreddegrad pr. århundrede
        baromega    %breddegrad af periapsis
        dbaromega   %ændring i breddegrad af periapsis pr. århundrede
        t           %tid
        b           %parameter
        c           %parameter
        s           %parameter
        f           %parameter
        trace
        trace_length=200
        ball
        text
    end

    methods
        function obj=Planet(name,color,planet_data)
            %Constructer
            if nargin>0
                obj.name=name;
                obj.color=color;
                obj.radius=planet_data(1);
                obj.a=planet_data(2);
                obj.da=planet_data(3);
                obj.e=planet_data(4);
                obj.de=planet_data(5);
                obj.L=planet_data(6);
                obj.dL=planet_data(7);
                obj.baromega=planet_data(8);
                obj.dbaromega=planet_data(9);
                obj.b=planet_data(10);
                obj.c=planet_data(11);
                obj.s=planet_data(12);
                obj.f=planet_data(13);
             end
        end




        function obj=update(obj,t)
            %Method for calculating the position of the planet at time t.
            if (t<-3000 || t>3000)
                error('t must be in [-3000,3000] for the model to be accurate')
            end
            obj.t=t;
            %Convert t to the correct time relative to the Julian Ephemeris
            %Date:
            t=obj.time_converter(t);
            a=obj.a+obj.da*t;
            e=obj.e+obj.de*t;
            L=obj.L+obj.dL*t;
            baromega=obj.baromega+obj.dbaromega*t;
            %We calculate M from the available data
            M=L-baromega+obj.b*t^2+obj.c*cosd(obj.f*t)+obj.s*sind(obj.f*t);
            M=obj.mod_M(M);
            estar=180/pi*e;
            %Our initial guess in Newtons method
            E_0=M+estar*sind(M);
            %We apply Newtons method to find E(t).
            E=newtons_method(E_0,@(E) E-estar*sind(E)-M, @(E) 1-e*cosd(E),10^(-6));
            %We calculate the coordinates of the planet at time t.
            obj.coordinates=[a*(cosd(E)-e), a*sqrt(1-e^2)*sind(E)]';
            if isempty(obj.ball)
                obj.make_planet();
            else
                obj.update_planet();
            end
        end

        function circle=make_circle(obj)
            t=linspace(0,2*pi,obj.resolution);
            circle=obj.coordinates+obj.radius*[cos(t);sin(t)];
        end

        function obj=make_planet(obj)
            circle=obj.make_circle();
            obj.ball=patch(circle(1,:),circle(2,:),obj.color,'visible','off');
            obj.text=text(obj.coordinates(1),obj.coordinates(2),obj.name);
            trace=obj.coordinates.*ones(2,obj.trace_length);
            obj.trace=plot(trace(1,:),trace(2,:));
        end

        function obj=update_planet(obj,d)
            circle=obj.make_circle();
            set(obj.ball, 'XData', circle(1,:),'YData',circle(2,:),'Visible','off')
            set(obj.text, 'Position', [obj.coordinates(1),obj.coordinates(2)]);
            set(obj.trace, 'XData', [obj.coordinates(1) obj.trace.XData(1:end-1)],'YData', [obj.coordinates(2) obj.trace.YData(1:end-1)]);
        end

        function t=time_converter(obj,t)
            %Method for converting t to Julian Ephemeris Date and then to
            %centuries past J2000.0.  This is because of the available data.
            t=(2816788-625674)*(t+3001)/(3000+3001)+625674;
            t= (t-2451545)/36525;
        end

        function M=mod_M(obj,M)
            %Convert M to a value between -180 deg and 180 deg.
            M=mod(M,360)-(mod(M,360)>=180)*360;
        end

    end
end

File: Workshop 3 Solsystemet-20230210/Python/Planet.py
import numpy as np
import matplotlib.pyplot as plt
from newtons_method import newtons_method

class Planet:
    # PLANET Class. Based on the method of https://ssd.jpl.nasa.gov/txt/aprx_pos_planets.pdf

    t = 0  # Tid
    coordinates = np.array([0, 0])
    resolution = 200
    trace_length = 200

    def __init__(self, name, color, planet_data, ax):
        # Constructer
        self.name = name
        self.color = color
        self.radius = planet_data[0]
        self.a = planet_data[1]  # Halve storakse
        self.da = planet_data[2]  # Ændring i halve storakse pr. århundrede
        self.e = planet_data[3]  # Excentricitet
        self.de = planet_data[4]  # Ændring i excentricitet pr. århundrede
        self.L = planet_data[5]  # Middelbreddegrad
        self.dL = planet_data[6]  # Ændring i middelbreddegrad pr. århundrede
        self.baromega = planet_data[7]  # Breddegrad af periapsis
        self.dbaromega = planet_data[8]  # Ændring i breddegrad af periapsis pr. århundrede
        self.b = planet_data[9]  # Parameter
        self.c = planet_data[10]  # Parameter
        self.s = planet_data[11]  # Parameter
        self.f = planet_data[12]  # Parameter
        self.ax = ax

    def update(self, t):
        # Method for calculating the position of the planet at time t.
        if t < -3000 or t > 3000:
            raise ValueError("t must be in [-3000, 3000] for the model to be accurate")
        self.t = t
        # Convert t to the correct time relative to the Julian Ephemeris Date:
        t = self.time_converter(t)
        a = self.a + self.da * t
        e = self.e + self.de * t
        L = self.L + self.dL * t
        baromega = self.baromega + self.dbaromega * t
        # We calculate M from the available data
        M = L - baromega + self.b * t ** 2 + self.c * np.cos(np.deg2rad(self.f * t)) + self.s * np.sin(
            np.deg2rad(self.f * t))
        M = self.mod_M(M)
        estar = np.deg2rad(e)
        # Our initial guess in Newtons method
        E_0 = M + estar * np.sin(M)
        # We apply Newtons method to find E(t).
        E = newtons_method(E_0, lambda E: E - estar * np.sin(E) - M, lambda E: 1 - e * np.cos(E), 1e-6)
        # We calculate the coordinates of the planet at time t.
        self.coordinates = np.array([a * (np.cos(E) - e), a * np.sqrt(1 - e ** 2) * np.sin(E)])
        if hasattr(self, "ball"):
            self.update_planet()
        else:
            self.make_planet()

    def make_circle(self):
        t = np.linspace(0, 2 * np.pi, self.resolution)
        circle = self.radius * np.array([np.cos(t), np.sin(t)])
        circle += self.coordinates[:, np.newaxis]
        return circle.T

    def make_planet(self):
        circle = self.make_circle()
        self.ball = plt.Polygon(circle, color=self.color)
        if self.ax: self.ax.add_patch(self.ball)
        self.text = plt.text(self.coordinates[0], self.coordinates[1], self.name)
        trace = np.tile(self.coordinates[:, np.newaxis], (1, self.trace_length))
        self.trace, = plt.plot(trace[0], trace[1])

    def update_planet(self):
        circle = self.make_circle()
        self.ball.set_xy(circle)
        self.text.set_position(self.coordinates)
        trace_x = [self.coordinates[0], *self.trace.get_xdata()[:-1]]
        trace_y = [self.coordinates[1], *self.trace.get_ydata()[:-1]]
        self.trace.set_data(trace_x, trace_y)

    def time_converter(self, t):
        # Method for converting t to Julian Ephemeris Date and then to
        # centuries past J2000 .0.This is because of the available data.
        t = (2816788 - 625674) * (t + 3001) / (3000 + 3001) + 625674
        t = (t - 2451545) / 36525
        return t

    def mod_M(self, M):
        # Convert M to a value between - 180 deg and 180 deg.
        M = np.mod(M, 360) - 360 * (np.mod(M, 360) >= 180)
        return M